File: scripts/r/mcp_server.R
# Source assessment skills module if running locally or inside package
if (file.exists("scripts/r/assessment_skills.R")) {
  source("scripts/r/assessment_skills.R")
} else if (file.exists("R/assessment_skills.R")) {
  source("R/assessment_skills.R")
}

#' Run MCP Server for NWFSC Data Catalog
#' @description Starts an MCP server exposing R SDK functions as tools alongside advanced biometrics skills.
#' @export
run_mcp_server <- function() {
  if (!requireNamespace("mcptools", quietly = TRUE)) {
    stop("The 'mcptools' package is required but not installed. Please install it first.")
  }
  if (!requireNamespace("ellmer", quietly = TRUE)) {
    stop("The 'ellmer' package is required but not installed. Please install it first.")
  }
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("The 'jsonlite' package is required but not installed. Please install it first.")
  }

  metadata_json <- get_sdk_metadata()
  metadata <- jsonlite::fromJSON(metadata_json, simplifyDataFrame = FALSE)

  tools <- list()

  # 1. Map standard SDK read_* functions into MCP tools
  for (item in metadata) {
    if (!startsWith(item$name, "read_")) {
      next
    }

    # Retrieve the exported function from our package namespace
    func <- getExportedValue("nwfscDataCatalog", item$name)

    # Build argument schema dynamically
    arguments <- list()
    for (param_name in names(item$parameters)) {
      if (param_name == "...") {
        next
      }
      param_desc <- item$parameters[[param_name]]
      if (param_name == "limit" || grepl("year", param_name)) {
        arguments[[param_name]] <- ellmer::type_integer(param_desc)
      } else {
        arguments[[param_name]] <- ellmer::type_string(param_desc)
      }
    }

    # Instantiate the ellmer tool
    t <- ellmer::tool(
      fun = func,
      name = item$name,
      description = item$description,
      arguments = arguments
    )

    tools[[length(tools) + 1]] <- t
  }

  # 2. Instantiate and append Advanced Fisheries Biometrics MCP Skills
  tool_expand_comps <- ellmer::tool(
    fun = skill_expand_compositions,
    name = "skill_expand_compositions",
    description = "Fisheries biometrics skill: Define latitude/depth strata and expand catch and biological/specimen data into length, age, and conditional age-at-length (CAAL) compositions. Returns file paths to the generated CSVs.",
    arguments = list(
      catch_data = ellmer::type_any("A data frame of catch records (with columns like year, catch_weight, lat, depth, etc.)."),
      bio_data = ellmer::type_any("A data frame of individual biological/specimen records (with columns like year, length_cm, age, sex, etc.).")
    )
  )

  tool_generate_index <- ellmer::tool(
    fun = skill_generate_index,
    name = "skill_generate_index",
    description = "Fisheries biometrics skill: Fit a spatiotemporal Tweedie sdmTMB model to catch data and extract a standardized index of abundance. Returns the path to the saved index CSV.",
    arguments = list(
      catch_data = ellmer::type_any("A data frame of catch/density records (must include spatial coordinates longitude/latitude or X/Y, and year)."),
      formula_str = ellmer::type_string("Formula string for the sdmTMB model (e.g. 'catch_weight ~ 0 + as.factor(year)').")
    )
  )

  tool_build_ss3 <- ellmer::tool(
    fun = skill_build_ss3,
    name = "skill_build_ss3",
    description = "Fisheries biometrics skill: Read a template Stock Synthesis (SS3) .dat file, inject new CPUE/index and composition blocks, and write out a new .dat file. Returns the path to the newly written .dat file.",
    arguments = list(
      index_csv = ellmer::type_string("Optional path to a CSV file containing the standardized abundance index to inject."),
      length_csv = ellmer::type_string("Optional path to a CSV file containing the expanded length compositions to inject."),
      age_csv = ellmer::type_string("Optional path to a CSV file containing the expanded marginal age compositions to inject."),
      template_dir = ellmer::type_string("Path to the directory containing template SS3 files (including a template .dat file).")
    )
  )

  tools[[length(tools) + 1]] <- tool_expand_comps
  tools[[length(tools) + 1]] <- tool_generate_index
  tools[[length(tools) + 1]] <- tool_build_ss3

  message("Starting NWFSC Data Catalog R MCP Server with advanced Biometrics Skills...")
  mcptools::mcp_server(tools = tools)
}

File: scripts/python/mcp_server.py
import json
import logging
import sys
from typing import Any, List, Optional, Union
try:
    from typing import Annotated
except ImportError:
    from typing_extensions import Annotated
from mcp.server.fastmcp import FastMCP
from nwfsc_data_catalog import ApiClient, Configuration, DataCatalog

# Configure logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
logger = logging.getLogger("nwfsc_data_catalog.mcp_server")

# Instantiate FastMCP server
mcp = FastMCP("NWFSC Data Catalog")

# Instantiate the Data Catalog SDK
config = Configuration()
api_client = ApiClient(config)
catalog = DataCatalog(api_client)

# Dynamically register all 'read_*' methods from catalog metadata
try:
    metadata_str = catalog.get_sdk_metadata()
    metadata = json.loads(metadata_str)
except Exception as e:
    logger.error(f"Failed to fetch SDK metadata from DataCatalog: {e}")
    sys.exit(1)

def simplify_type(type_str: str) -> str:
    type_str_lower = type_str.lower()
    if "list" in type_str_lower:
        return "Optional[List[str]]"
    elif "bool" in type_str_lower:
        return "Optional[bool]"
    elif "int" in type_str_lower:
        return "Optional[int]"
    elif "float" in type_str_lower or "number" in type_str_lower:
        return "Optional[float]"
    elif "str" in type_str_lower:
        return "Optional[str]"
    else:
        return "Optional[Any]"


dynamic_funcs = {}

for item in metadata:
    name = item.get("name")
    if not name or not name.startswith("read_"):
        continue

    description = item.get("description", "Fetch data from the NOAA NWFSC Data Catalog.")
    parameters = item.get("parameters", {})

    # Build function signature arguments and call arguments
    sig_args = []
    call_args = []

    for p_name, p_type in parameters.items():
        if p_name in ("self", "kwargs", "**kwargs") or p_name.startswith("_"):
            continue

        clean_type = simplify_type(p_type)
        sig_args.append(f"{p_name}: {clean_type} = None")
        call_args.append(f"{p_name}={p_name}")

    # Add standard dynamic parameters (limit and fields) if not already declared in signature
    if "limit" not in parameters:
        sig_args.append("limit: int = 1000")
        call_args.append("limit=limit")
    if "fields" not in parameters:
        sig_args.append("fields: Optional[List[str]] = None")
        call_args.append("fields=fields")

    sig_str = ", ".join(sig_args)
    call_str = ", ".join(call_args)

    # Dynamic function definition
    fn_code = f"""
def {name}({sig_str}) -> Any:
    \"\"\"{description}\"\"\"
    return catalog.{name}({call_str})
"""

    ctx = {
        "catalog": catalog,
        "Any": Any,
        "List": List,
        "Optional": Optional,
        "Union": Union,
        "Annotated": Annotated,
    }

    try:
        exec(fn_code, ctx)
        fn = ctx[name]
        # Keep reference to prevent GC
        dynamic_funcs[name] = fn
        # Register tool with FastMCP
        mcp.tool(name=name, description=description)(fn)
        logger.info(f"Registered dynamic tool: {name}")
    except Exception as e:
        logger.error(f"Failed to dynamically define/register tool {name}: {e}")


def main():
    """Runs the FastMCP server."""
    # Register assessment skills
    try:
        from nwfsc_data_catalog.assessment_skills import (
            skill_generate_design_index, 
            skill_expand_compositions, 
            skill_build_ss3
        )
        mcp.tool(name="skill_generate_design_index", description="Calculate stratified area-swept biomass indices, SE, and CV across depth/latitude strata.")(skill_generate_design_index)
        mcp.tool(name="skill_expand_compositions", description="Perform weighted stratum expansions on length and age frequencies.")(skill_expand_compositions)
        mcp.tool(name="skill_build_ss3", description="Parse an SS3 .dat template file, inject new index/composition data, and write out a new file.")(skill_build_ss3)
        logger.info("Registered assessment skill tools.")
    except ImportError as e:
        logger.warning(f"Could not import assessment skills (this is normal during bootstrapping): {e}")

    logger.info("Starting NWFSC Data Catalog FastMCP Server...")
    mcp.run()


if __name__ == "__main__":
    main()
